### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by F4ever**

### CODE REVIEW

File: src/transport/msg_schemas.py [hunk 1]
```diff
@@ -131,6 +143,11 @@ def check_pause_message(msg: PauseMessage) -> bool:
             ),
         )
 
+        if not verified:
+            logger.error({'msg': 'Message verification failed.', 'value': msg})
```
Comment (avsetsin) [inline src/transport/msg_schemas.py:147]: I'd also like to have a metric and an alert for this scenario

File: src/transport/msg_schemas.py [hunk 2]
```diff
@@ -118,10 +130,10 @@ class PauseMessage(TypedDict):
     stakingModuleId: int
 
 
-def get_pause_messages_sign_filter(pause_prefix: str) -> Callable:
+def get_pause_messages_sign_filter(attestation_prefix: bytes) -> Callable:
```
Comment (avsetsin) [inline src/transport/msg_schemas.py:133]: why attestation_prefix?

File: src/metrics/metrics.py [hunk 3]
```diff
@@ -13,32 +12,31 @@
...
     'Guardians deposit messages',
     ['address', 'version'],
     namespace=PREFIX,
 )
-KAFKA_PAUSE_MESSAGES = Gauge(
-    'kafka_pause_messages',
+PAUSE_MESSAGES = Gauge(
+    'pause_messages',
     'Guardians pause messages',
     ['address', 'version'],
```
Comment (avsetsin) [inline src/metrics/metrics.py:35]: I think it's worth adding a label `module_id` for pause and deposit messages so we can separate them on the dashboard and have the details in the alert

File: src/bots/pause.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        if not self.w3.lido.staking_router.is_staking_module_active(module_id):
+            # Module already deactivated
+            self._clear_outdated_messages_for_module(module_id)
+            logger.info({'msg': f'Module {module_id} already paused. Skip message.'})
+            return False
+
+        pause_tx = self.w3.lido.deposit_security_module.pause_deposits(
+            message['blockNumber'],
+            module_id,
+            (message['signature']['r'], compute_vs(message['signature']['v'], message['signature']['s']))
```
Comment (avsetsin) [inline src/bots/pause.py:96]: can use `_vs`

File: src/depositor.py [hunk 5]
```diff
@@ -1,48 +1,74 @@
...
     depositor_bot = DepositorBot(w3)
-    depositor_bot.run_as_daemon()
+
+    e = Executor(
+        w3,
+        depositor_bot.execute,
+        5,
+        variables.MAX_CYCLE_LIFETIME_IN_SECONDS,
+    )
+    logger.info({'msg': 'Rum executor.'})
```
Comment (avsetsin) [inline src/depositor.py:70]: 🥃

File: src/pauser.py [hunk 6]
```diff
@@ -1,37 +1,67 @@
...
-    deposit_pause_bot = PauserBot(w3)
-    deposit_pause_bot.run_as_daemon()
+    pause = PauserBot(w3)
+    e = Executor(
+        w3,
+        pause.execute,
+        1,
+        variables.MAX_CYCLE_LIFETIME_IN_SECONDS,
+    )
+    logger.info({'msg': 'Rum executor.'})
```
Comment (avsetsin) [inline src/pauser.py:63]: 🥃

File: src/bots/pause.py
Comment (avsetsin) [inline src/bots/pause.py:1]: pause.py -> pauser.py?

there's "pause" in some places and "pauser" in others

File: README.md [hunk 7]
```diff
@@ -1,92 +1,160 @@
...
+| WEB3_RPC_ENDPOINTS        | -                                          | List of rpc endpoints that will be used to send requests comma separated (`,`)                                           |
+| WALLET_PRIVATE_KEY        | -                                          | Account private key                                                                                                      |
+| CREATE_TRANSACTIONS       | false                                      | If true then tx will be send to blockchain                                                                               |
+| LIDO_LOCATOR              | 0xC1d0b3DE6792Bf6b4b37EccdcC24e45978Cfd2Eb | Lido Locator address. Mainnet by default. Other networks could be found [here](https://docs.lido.fi/deployed-contracts/) |
+| DEPOSIT_CONTRACT          | 0x00000000219ab540356cBB839Cbe05303d7705Fa | Ethereum deposit contract address                                                                                        |
+| DEPOSIT_MODULES_WHITELIST | -                                          | List of staking module's ids in which the depositor bot will make deposits                                               |
```
Comment (avsetsin) [inline README.md:50]: looks like it has default value
```python
DEPOSIT_MODULES_WHITELIST = [int(module_id) for module_id in os.getenv('DEPOSIT_MODULES_WHITELIST', '1').split(',')]
```

File: README.md [hunk 8]
```diff
@@ -1,92 +1,160 @@
...
+- [Annotations to code](#annotations-to-code)
+
+## Running Daemon
+
+1. Create `.env` file 
+2. Setup variables
+   - Set WEB3_RPC_ENDPOINTS
+   - Set WALLET_PRIVATE_KEY
+   - Set CREATE_TRANSACTIONS to true
+   - Set TRANSPORTS to rabbit
```
[the remark this answers is not in the document]
Comment (F4ever) [reply README.md:33]: fixed

File: docker-compose.yml [hunk 9]
```diff
@@ -24,10 +28,11 @@ services:
...
     environment:
-      - NETWORK=${NETWORK}
+      - WEB3_RPC_ENDPOINTS=${WEB3_RPC_ENDPOINTS}
       - WALLET_PRIVATE_KEY=${WALLET_PRIVATE_KEY}
-      - KAFKA_BROKER_ADDRESS_1=${KAFKA_BROKER_ADDRESS_1}
-      - KAFKA_USERNAME=${KAFKA_USERNAME}
-      - KAFKA_PASSWORD=${KAFKA_PASSWORD}
-      - KAFKA_TOPIC=${KAFKA_TOPIC}
+      - CREATE_TRANSACTIONS=${CREATE_TRANSACTIONS}
+      - TRANSPORTS=${TRANSPORTS}
```
[the remark this answers is not in the document]
Comment (F4ever) [reply docker-compose.yml:34]: fixed

File: src/blockchain/contracts/lido.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+from blockchain.contracts.base_interface import ContractInterface
+
+
+logger = logging.getLogger(__name__)
+
+
+class LidoContract(ContractInterface):
+    abi_path = './interfaces/Lido.json'
+
+    def get_depositable_ether(self, block_identifier: BlockIdentifier = 'latest') -> Wei:
```
[the remark this answers is not in the document]
Comment (F4ever) [reply src/blockchain/contracts/lido.py:14]: Yes, but if I'm going to write a library with an autogenerating class, there will be block_identifier)

File: src/blockchain/deposit_strategy/curated_module.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    def _calculate_recommended_gas_based_on_deposit_amount(self, deposits_amount: int) -> Wei:
+        # For one key recommended gas fee will be around 10
+        # For 10 keys around 100 gwei. For 20 keys ~ 800 gwei
+        # ToDo percentiles for all modules?
+        recommended_max_gas = (deposits_amount ** 3 + 100) * 10 ** 8
+        logger.info({'msg': 'Calculate recommended max gas based on possible deposits.'})
+        GAS_FEE.labels('based_on_buffer_fee', self.module_id).set(recommended_max_gas)
+        return recommended_max_gas
+
+    def _get_pending_base_fee(self):
```
[the remark this answers is not in the document]
Comment (F4ever) [reply src/blockchain/deposit_strategy/curated_module.py:64]: Fixed

File: src/blockchain/deposit_strategy/curated_module.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+        GAS_FEE.labels('max_fee', self.module_id).set(variables.MAX_GAS_FEE)
+
+        current_buffered_ether = self.w3.lido.lido.get_depositable_ether()
+        if current_buffered_ether > variables.MAX_BUFFERED_ETHERS:
+            return variables.MAX_GAS_FEE >= current_gas_fee
+
+        return recommended_gas_fee >= current_gas_fee
+
+    def _get_recommended_gas_fee(self):
```
[the remark this answers is not in the document]
Comment (F4ever) [reply src/blockchain/deposit_strategy/curated_module.py:84]: fixed

File: src/bots/depositor.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+            b'',
+            signs,
+        )
+
+    @staticmethod
+    def _prepare_signs_for_deposit(quorum: list[DepositMessage]) -> tuple[tuple[str, str]]:
+        sorted_messages = sorted(quorum, key=lambda msg: int(msg['guardianAddress'], 16))
+
+        return tuple(
+            (msg['signature']['r'], compute_vs(msg['signature']['v'], msg['signature']['s']))
```
[the remark this answers is not in the document]
Comment (F4ever) [reply src/bots/depositor.py:206]: I use r, s, v signature to verify message and _vs stays unverified. Thats why I don't use it

File: src/depositor.py (hunk 5 above)
Comment (F4ever) [reply src/depositor.py:70]: 🥃

File: src/bots/depositor.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+        """Returns True if module is ready for deposit"""
+        return self.w3.lido.staking_router.is_staking_module_active(module_id)
+
+    def _get_quorum(self, module_id: int) -> Optional[list[DepositMessage]]:
+        """Returns quorum messages or None is quorum is not ready"""
+        actualize_filter = self._get_message_actualize_filter(module_id)
+        messages = self.message_storage.get_messages(actualize_filter)
+        min_signs_to_deposit = self.w3.lido.deposit_security_module.get_guardian_quorum()
+
+        CURRENT_QUORUM_SIZE.labels('required').set(min_signs_to_deposit)
```
[the remark this answers is not in the document]
Comment (avsetsin) [reply src/bots/depositor.py:136]: I mean, we should have that label, but we don't. Quorum is different for each module

File: src/bots/pause.py (hunk 4 above)
Comment (F4ever) [reply src/bots/pause.py]: https://github.com/lidofinance/depositor-bot/pull/170#discussion_r1332763513

File: src/bots/depositor.py [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+
+        gas_is_ok = module_strategy.is_gas_price_ok()
+        logger.info({'msg': 'Calculate gas recommendations.', 'value': gas_is_ok})
+
+        keys_amount_is_profitable = module_strategy.is_deposited_keys_amount_ok()
+        logger.info({'msg': 'Calculations deposit recommendations.', 'value': keys_amount_is_profitable})
+
+        if is_depositable and quorum and gas_is_ok and keys_amount_is_profitable:
+            logger.info({'msg': 'Checks passed. Prepare deposit tx.'})
+            return self._build_and_send_deposit_tx(quorum)
```
[the remark this answers is not in the document]
Comment (F4ever) [reply src/bots/depositor.py:118]: Can add it in future versions, but not sure it is very usefull

**Review by infloop — APPROVED** [review]

**Pull request merged**